fendermint/vm/interpreter/src/fvm/state/snapshot.rs

```diff
@@ -0,0 +1,312 @@
+    pub async fn write_car(self, path: impl AsRef<Path>) -> anyhow::Result<()> {
+        let file = tokio::fs::File::create(path).await?;
+
+        // derive the car header roots
+        let car = CarHeader::new(self.car_header_roots()?, self.version());
+
+        match self {
+            Snapshot::V1(s) => s.write_car(car, file).await?,
+        }
+
+        Ok(())
+    }
+
+    fn car_header_roots(&self) -> anyhow::Result<Vec<Cid>> {
+        match self {
+            Snapshot::V1(s) => s.car_header_roots(),
+        }
+    }
+}
+
+pub struct V1Snapshot<DB> {
+    /// The root cid of the state tree
+    state_tree_root: Cid,
+    /// The state tree of the current blockchain
+    state_tree: StateTree<ReadOnlyBlockstore<DB>>,
+    /// The latest block height with state params serialized to bytes
+    block_state_params_bytes: Vec<u8>,
+    /// Block state params cid derived from block state params bytes
+    block_state_params_cid: Cid,
+}
+
+impl<DB> V1Snapshot<DB>
+where
+    DB: Blockstore + Clone + 'static + Send,
+{
+    /// Creates a new V2Snapshot struct. Caller ensure store
+    pub fn new(
+        store: DB,
+        state_params: FvmStateParams,
+        block_height: BlockHeight,
+    ) -> anyhow::Result<Self> {
+        let state_tree =
+            StateTree::new_from_root(ReadOnlyBlockstore::new(store), &state_params.state_root)?;
+        let state_tree_root = state_params.state_root;
+
+        let block_state_params_bytes = fvm_ipld_encoding::to_vec(&(state_params, block_height))?;
+        let block_state_params_cid =
+            Cid::new_v1(DAG_CBOR, Code::Blake2b256.digest(&block_state_params_bytes));
+
+        Ok(Self {
+            state_tree_root,
+            state_tree,
+            block_state_params_bytes,
+            block_state_params_cid,
+        })
+    }
+
+    /// For V1 snapshot, we are putting two components into the CAR file: the state tree and latest
+    /// state params. One header is for the state tree root. The other header root cid is for the
+    /// serialized block state params.
+    fn car_header_roots(&self) -> anyhow::Result<Vec<Cid>> {
+        Ok(vec![self.state_tree_root, self.block_state_params_cid])
+    }
+
+    pub fn init_chain(&self) -> anyhow::Result<()> {
+        todo!()
+    }
+
+    async fn write_car(self, car: CarHeader, write: tokio::fs::File) -> anyhow::Result<()> {
+        let write_task = tokio::spawn(async move {
+            let mut streamer =
+                StateTreeStreamer::new(self.state_tree_root, self.state_tree.into_store());
+            let mut write = write.compat_write();
+
+            car.write_stream_async(&mut Pin::new(&mut write), &mut streamer)
+                .await
+                .unwrap();
+
+            let mut streamer = tokio_stream::iter(vec![(
+                self.block_state_params_cid,
+                self.block_state_params_bytes,
+            )]);
+            car.write_stream_async(&mut Pin::new(&mut write), &mut streamer)
```

Comment:
According to the docs, `write_strea_async` writes the header and stream of data to writer. Would calling it twice write the header twice? We should double-check this to ensure that we are not writing the header twice.

If you look at the CAR spec, there is only one header with the roots at the beginning of the file: https://ipld.io/specs/transport/car/carv2/

Reply:
yeah, it's only written once now